artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/impl/PagingStoreImpl.java

```diff
@@ -1262,8 +1324,16 @@ private boolean writePage(Message message,
                              Transaction tx,
                              RouteContextList listCtx,
                              Function<Message, Message> pageDecorator) throws Exception {
-      lock.writeLock().lock();
-
+      if (timedWriter == null) {
+         // in case timedWriter is not being used, we need to guarantee a writeLock.
+         // because there's no way to upgrade from reading to writeLock
+         writeLock();
+      } else {
+         // if we are using a timedWritter, we can just use a readLock to guarantee paging is not changed
+         // this will issue the writer on a different thread, and we can just use a readLock
+         readLock();
```

Comment:
I dont understand this. When a timedWriter is around it _doesnt obviously take the write lock anywhere_, and if it doesnt do so then how does taking the readLock here stop it changing paging? The only thing obviously taking the writeLock when adding new stuff is this writePage method, _if timedWriter is null_.

What am I missing?

Reply:
I will improve the comment there.


We need to keep a lock while we checked for isPaging... we need to make sure the system is paging for the entire duration of the method, hence the readLock.


When timedWriter is null (which will only happen in tests), I still  need to hold the lock, however later on when moving to a different page a writeLock will be needed, and that would break things with a deadlock.

Reply:
Ok, so the comment wasnt saying "paging is not changed" in the sense that things are added/removed/updated from paging, but is more specifically saying "the 'paging' field is not altered"? The [updated] comment then at least makes sense, since the 'paging' field is seemingly only directly updated in cases that do hold the write lock.

I still dont really get that timedWriter, the thing writing pages if its around, isnt ever holding the write lock, when other things that also update the pages and run on different threads, do hold it. Surely that means those two things can happen concurrently with each other? Plus that timedWriter can also update things concurrent with stuff that hold the read or write lock?

Reply:
thing is... when the pagedTimeWriter is in use, things are pretty much single threaded.. and updated from the PageExecutor. The only difference is to schedule a task it needs to use a readLock to make sure it was paging, and it should stay paging until the task is scheduled.